File: lib/protocol/protocol.cpp
#include <Arduino.h>
#include "protocol.h"

#define BAUDRATE 115200

bool protocol_init(void) 
{
    Serial.begin(BAUDRATE);
    return Serial; // Check if serial port is initialized
}

bool protocol_send(const uint8_t *data, size_t dlen) 
{
    return (dlen == Serial.write(data, dlen)); // Return true if all bytes were sent successfully
}

size_t protocol_receive(uint8_t *buffer, size_t blen) 
{
    while (0 == Serial.available())
    {
        ;   // Wait until data is available
    }
    return Serial.readBytes(buffer, blen); // Return the number of bytes read
}

File: client/protocol.py
import serial
import serial.tools.list_ports

ser: serial.Serial  # Serial connection object

def enumerate_serial_ports():
    return [port.device for port in serial.tools.list_ports.comports()]

def protocol_init(port, baudrate=115200) -> bool:
    global ser
    ser = serial.Serial(port, baudrate)
    return ser.is_open

def protocol_send(data) -> bool:
    global ser
    ret = False
    if ser.is_open == True:
        ret = (len(data) == ser.write(data))  # Send
    return ret
    

def protocol_receive(size: int):
    global ser
    data = bytes()
    if ser.is_open == True:
        data = ser.read(size)
    return data
